server/src/main/java/org/elasticsearch/search/aggregations/metrics/GlobalOrdCardinalityAggregator.java

```diff
@@ -64,12 +90,182 @@ public GlobalOrdCardinalityAggregator(
+                while (top.docID() < target) {
+                    top.advance(target);
+                    top = nonVisitedPostings.updateTop();
+                }
+                return doc = top.docID();
+            }
+        }
+
+        @Override
+        public long cost() {
+            return docsWithField.cost();
+        }
+
+        void startPruning() throws IOException {
+            dynamicPruningSuccess++;
+            nonVisitedOrds = new HashMap<>();
+            // TODO: iterate the bitset using a `nextClearBit` operation?
+            for (long ord = 0; ord < maxOrd; ++ord) {
+                if (visitedOrds.get(ord)) {
+                    continue;
+                }
+                BytesRef term = values.lookupOrd(ord);
+                if (indexTerms.seekExact(term) == false) {
+                    // This global ordinal maps to a value that doesn't exist in this segment
+                    continue;
+                }
+                nonVisitedOrds.put(ord, indexTerms.postings(null, PostingsEnum.NONE));
+            }
+            nonVisitedPostings = new PriorityQueue<>(nonVisitedOrds.size()) {
+                @Override
+                protected boolean lessThan(PostingsEnum a, PostingsEnum b) {
+                    return a.docID() < b.docID();
+                }
+            };
+            for (PostingsEnum pe : nonVisitedOrds.values()) {
+                nonVisitedPostings.add(pe);
+            }
+        }
+
+        void onVisitedOrdinal(long ordinal) throws IOException {
+            numNonVisitedOrds--;
+            if (nonVisitedOrds == null) {
+                if (numNonVisitedOrds <= MAX_TERMS_FOR_DYNAMIC_PRUNING) {
+                    startPruning();
+                }
+            } else {
+                if (nonVisitedOrds.remove(ordinal) != null) {
+                    // Could we make this more efficient?
+                    nonVisitedPostings.clear();
+                    for (PostingsEnum pe : nonVisitedOrds.values()) {
+                        nonVisitedPostings.add(pe);
+                    }
+                }
+            }
+        }
     }
 
     @Override
     public LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCtx, final LeafBucketCollector sub) throws IOException {
         values = valuesSource.globalOrdinalsValues(aggCtx.getLeafReaderContext());
+
+        if (parent == null && field != null) {
+            // This optimization only applies to top-level cardinality aggregations that apply to fields indexed with an inverted index.
+            final Terms indexTerms = aggCtx.getLeafReaderContext().reader().terms(field);
+            if (indexTerms != null) {
+                BitArray bits = visitedOrds.get(0);
+                final int numNonVisitedOrds = maxOrd - (bits == null ? 0 : (int) bits.cardinality());
+                if (maxOrd <= MAX_FIELD_CARDINALITY_FOR_DYNAMIC_PRUNING || numNonVisitedOrds <= MAX_TERMS_FOR_DYNAMIC_PRUNING) {
```

Comment:
So the optimisation also kicks in on fields with more than 1024 unique values, if there 128 or less terms to be processed. The brute force leaf bucket collector implementation does update the bit sets used here to determine `numNonVisitedOrds`.